src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1006,8 +1006,9 @@ internal async Task Cleanup(bool async, bool connectionClosing = false, bool isD
                 Log.Debug($"Query duration time: {Connector.QueryLogStopWatch.ElapsedMilliseconds}ms", Connector.Id);
                 Connector.QueryLogStopWatch.Reset();
             }
-            Connector.EndUserAction();
+            Command.CommandStop();
```

Comment:
I suddenly thought of something... It isn't clear to me whether the activity should end when the reader is disposed (as this PR does now), or when the reader is first returned to the user. The problem with the former is that the time includes application processing time, which doesn't seem very relevant... In the extreme case, the database returns all results immediately, but for some reason the application processes them much later, inflating the duration in a way that doesn't seem productive.

On the other hand, if PG is returning a large resultset, then considerable database computation/work can occur after the first row has been returned (i.e. after the reader has been returned to the application).

I don't think there's a 100% "correct" answer here - the two definitions each has their advantages/disadvantages. But I think the former one (time to first response) is probably what most people want. I've made a note to bring this up when discussing the database draft (I hope to start getting involved later this year).

(if we do "time to first response", the implementation probably becomes simpler, and the activity is no longer correlated to the user action)

Reply:
The activity is ending when the reader is disposed because that's how we've logged it until now. If we do change activity tracing, the same thing should be done for logging.

Reply:
There are downsides to both. As an application author (and telemetry consumer) I usually want to see the whole period for which the `DbDataReader` is alive captured as the span. (IME, the time is usually spent deserialising the results, which is relevant to capture. I do understand that in some cases, application code does "lazily" enumerate the results and intermingle it with complex processing, which might not directly reflect the DB processing time. OTOH, it's still consuming a DB connection the entire time, which seems helpful to capture in telemetry.)

At https://github.com/mysql-net/MySqlConnector/issues/1036 I suggested adding an event for when the `DbDataReader` is created; this would let telemetry systems divide the span into two parts.

Reply:
> OTOH, it's still consuming a DB connection the entire time, which seems helpful to capture in telemetry.)

It's going to consume a DB connection for as long as the connection is not closed/returned to the pool, so I'm not sure how exactly it's going to help you.

Reply:
Yes, that's a good point; there's no guarantee that the connection is disposed quickly following the reader being disposed. (Although I think it is a common pattern to write: `using var connection = ...; using var command = ...; using var reader = command.ExecuteReader(); ...`.)

Reply:
This is a thorny issue which we also discussed offline a bit. We used the terminology TTFR (Time Till First Row) and TTLR (Time Till Last Row) to refer to both (even if TTLR should technically refer to disposing the reader, which isn't quite the same). Some summary and additional thoughts:

* I'm not sure about the exact downsides of doing both (i.e. having a TTFR subspan within the TTLR span). Visualization-wise, if by default you see the two spans, then I think it could be confusing and therefore detrimental to most users, who don't care about this distinction; we don't want to burden the vast majority of users with low-level high-resolution data which only some users are going to care about or even understand. So I do think it may be better to pick one now rather than doing both.
* Outside of the .NET world, many drivers buffer entire resultsets by default rather than stream; in that case, it seems pretty clear that TTLR is the right choice, without including any application processing that comes after buffering is complete. 
* An advantage of doing TTLR is that we can always add a TTFR subspan later without breaking anyone; the opposite isn't true.

It feels like we're converging towards TTLR (though @vonzshik talked to some people with TTFR tendencies). It's also worth remembering everything is experimental/draft, so we can also release an implementation with a clear experimental status etc.

/cc @NinoFloris who I think was in the TTLR camp

Reply:
Definitely in favour of TTLR. (I'd probably leave it up to each ADO.NET driver to decide if that's `DbDataReader.Dispose` or earlier if it can easily be detected that the last row was read from the network. But I'm also happy with an "official" ADO.NET recommendation for one or the other.)

As mentioned earlier, can we get the best of both worlds (without nested spans) by making the whole span TTLR and calling `AddEvent` for TTFR (within that one span)?

Reply:
Does the concept of an event within a span exist in OpenTelemetry (have I missed it)? Or would this be a DiagnosticSource thing that doesn't map to OpenTelemetry? If it's the latter, we certainly can go down that path, though I'm really not sure how valuable it is to build/standardize non-OTel stuff now that OTel is the cool new standard...

(nested spans may not be that bad - I'm honestly not sure)

Reply:
Yes: "Spans encapsulate: A list of timestamped Events"

https://github.com/open-telemetry/opentelemetry-specification/blob/2f23c2013ff49cf45de152019367d0ec0f2135d1/specification/trace/api.md#span

They're used for exceptions: "An exception SHOULD be recorded as an Event on the span during which it occurred."

https://github.com/open-telemetry/opentelemetry-specification/blob/2f23c2013ff49cf45de152019367d0ec0f2135d1/specification/trace/semantic_conventions/exceptions.md#recording-an-exception

Reply:
Then that sounds pretty good to me!

Reply:
As another example, someone might want to submit `Activity` objects to Application Insights (via [TrackDependency](https://docs.microsoft.com/en-us/dotnet/api/microsoft.applicationinsights.telemetryclient.trackdependency?view=azure-dotnet)). Since that's a different "sink", you'd want to follow AppInsights conventions for naming. Again, having a standardised list of ADO.NET Activity names would help that bridge subscribe to the desired events and submit them properly.

For AppInsights, the call might look like this: `TrackDependency("MySQL", $"{conn.DataSource} | {conn.Database}", "SELECT", command.CommandText);` (so the attributes are slightly different than what OTel wants to receive).

We shouldn't unnecessarily use OpenTelemetry names that different bridges might have to "undo" in order to submit this (platform neutral) `ActivitySource` data to other sinks.